WIP autoencoder

beckermr opened the pull request:
Thus PR adds an autoencoder to muffnn. The autoencoder supports a variety of metrics and is compatible with the scikit-learn transform interface. 

I additionally 

1. refactored the MLP classes into their own subpackage.
2. pinned conda to 4.0.5 to avoid conda bugs
3. upgraded TensorFlow to 0.11.0rc2

Closes #8

beckermr commented:
@mheilman See the code in this PR for my first pass. I am still shaking the bugs out, but the MSE metric only works great.

beckermr commented:
This is a conda bug. :(

beckermr commented:
@mheilman Thus PR is ready. Unfortunately it ballooned to ~1500 after I wrote all of the tests. :(

I suggest we do it in chunks:

1. The refactoring of the MLP and base classes and the Dockerfile.
2. The autoencoder class itself.
3. The autoencoder tests.

beckermr commented:
I am going to add a test for sparse inputs because why not. :)

mheilman commented:
sorry for the slow response on this. i'm at a conference, so i probably won't be able to take a close look until next week.

beckermr commented:
No worries! I am continuing to hack away on this in the mean time.

beckermr commented:
@mheilman I rewrote the tests to eliminate a several hundred lines of duplicate code and removed a bunch of tests that duplicated each other for the most part.

mheilman commented on muffnn/__init__.py:
```diff
@@ -1,5 +1,6 @@
+from .autoencoder.autoencoder import Autoencoder
+from .core import TFPicklingBase
 
-__all__ = ['MLPBaseEstimator', 'MLPClassifier', 'MLPRegressor']
```
fwiw, i don't think you need `muffnn/autoencoder/__init__.py`

beckermr replied:
So after python 3.2, removing this file is OK I think. From the setuptools docs

"Anyway, find_packages() walks the target directory, filtering by inclusion patterns, and finds Python packages (any directory). On Python 3.2 and earlier, packages are only recognized if they include an __init__.py file. Finally, exclusion patterns are applied to remove matching packages."

I think we might want to retain it for all python 3 versions?

mheilman commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -0,0 +1,682 @@
+from sklearn.utils import check_array, check_random_state
+from sklearn.utils.validation import NotFittedError
+
+from ..core import TFPicklingBase, affine
```
I think it's best to avoid relative imports and just to `from muffnn.core ...` here.

beckermr replied:
Not following why...

mheilman replied:
it doesn't really matter much, but "explicit is better than implicit" and all that.  absolute imports are easier for the reader, imo

beckermr replied:
hmmmm. I actually don't know this part of the python interpreter. Suppose I have two copies of muffnn, maybe one installed and another that is being run directly from a local dir. With an absolute import, which copy gets used?

mheilman replied:
hmm... i *think* this issue is orthogonal to how packages get loaded from the filesystem.  iirc, if you're in a directory with a subdirectory that's a package, that'll take priority regardless of relative or absolute importing.

mheilman commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -0,0 +1,682 @@
+        autoencoder. For example, if hidden_units is (256, 32, 16) then
+        the autoencoder has layers
+
+            [input_layer, 256, 32, 16, 32, 256, input_layer]
```
it seems like the second `input_layer` should be `output_layer`, or perhaps better yet, that both `input_layer`s should be `input_layer_size`.

mheilman commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -0,0 +1,682 @@
+    activation : (callable, callable), optional
+        The activation functions for
+
+            (the bulk of the autoencoder and encoding layer, output layer).
```
unfinished here?

mheilman replied:
at a second glance, i see what this means.  I think it'd be cleared to say "A tuple of two callables for 1) the bulk of the autoencoder and encoding layer, and 2) the output layer."

mheilman commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -0,0 +1,682 @@
+        used.
+    learning_rate : float, optional
+        Learning rate for Adam.
+    metric : string, optional
```
i'd suggest either a) removing this kwarg or b) letting the user specify metrics for discrete variables in addition to ordinal variables.

beckermr replied:
What metrics besides cross-entropy w/ sigmoid/softmax do you expect to be useful for discrete variables?

mheilman replied:
i'm not sure if it really makes sense for autoencoders, but perhaps hinge loss (SVMs) or other loss functions that sklearn provides as [options for its SGD classifiers](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html).

mheilman replied:
i think i understand your approach here, and i think it makes sense, but it seems like maybe there's a better way to expose the options.  how about this?

```
"""
An autoencoder using TensorFlow for discrete, continuous, or mixed data.

Parameters
----------
....
hidden_activation : callable, optional
    The activation function for the hidden layers.
    See `tensorflow.nn` for various options.
output_activation : callable, optional
    The activation function for the output layer.
    See `tensorflow.nn` for various options.
...
metric : string, optional
    Default metric for the autoencoder. Options are
        'mse' - mean square error
        'cross-entropy' - cross-entropy
    Note that this will be overridden for columns specified by `binary_indices`
    or `categorical_indices`
binary_indices : array-like, shape (n_binary,), optional
    Array of indices for which `tf.nn.sigmoid` will be used for the
    output layer activation and cross-entropy will be used in the loss
    function.
categorical_indices : array-like, shape (n_categorical, 2), optional
    An array where each row specifies a range of indices for a categorical
    variable that has been expanded to multiple indices (e.g., 1-hot encoded).
    The first column contains start indices. The second column contains
    lengths. For each range of indices, the output layer will use
    `tensorflow.nn.softmax` activation and the loss function will use
    cross-entropy.
...
"""
```

also, I don't think cross-entropy should force a sigmoid output layer.  One might want to use either softmax or sigmoid activation with cross-entropy, I think, depending on the problem.

beckermr replied:
This looks good to me. I will adjust the code accordingly.

mheilman commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -0,0 +1,682 @@
+        Array of indices for which `tf.nn.sigmoid` will be used for the
+        output layer activation and the cross-entropy will be used as the
+        metric for the autoencoder.
+    categorical_begin : array-like, shape (n_categorial), optional
```
it seems like it'd be easier to just have one kwarg for categorical variables, of shape `(n_categorical, 2)` (there's a typo in docstring btw).

beckermr replied:
Sure I can make this change. I like the current structure since it is more explicit and you don't have to remember the order (i.e. is it `(begin, size)` or `(size, begin)`?)

mheilman replied:
`(begin, size)` seems like it'd make sense.  offset, length is pretty common (e.g., SQL).  one could also do `(begin, end)`.

It seems unusual to have 2 keyword arguments that are optional but require each other, particular when they have the same data type.

mheilman commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -0,0 +1,682 @@
+    dropout : float or None, optional
+        The dropout probability. If None, then dropout will not be used.
+        Note that dropout is not applied to the input layer.
+    activation : (callable, callable), optional
```
related to the comment below about `metric`, i'd suggest having this argument just be for the hidden layer.

With respect to the output layer, I'd suggest either a) just not letting the user specifying the output activation for now, or b) having a separate kwarg that would allow specification of activations for discrete as well as ordinal variables.

I'd go for the (a) and maybe add that functionality later if it's really needed.

sigmoid/softmax seems more obviously correct for discrete variables and less likely to need changing than MSE for ordinal/continuous data, of course, but it seems strange to have some the activation set in stone for some types of data but not others.

mheilman commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -0,0 +1,682 @@
+                                                      feed_dict=feed_dict))
+        return np.concatenate(pred_batches)
+
+    def inverse_transform(self, X, y=None, return_sparse=False):
```
What's the use case for this?

beckermr replied:
If people do modeling on the encoded variables (say fitting a density estimate) that produces new ones, they can transform back to the input space.

beckermr replied:
See the KDE digits example from scikit for a typical use case.

mheilman replied:
ah ok.  it might be worth adding a sentence explaining the use case.  i think most users wouldn't need this and would be a bit confused by it

mheilman commented:
lemme know whenever you're ready for me to take another look

beckermr commented:
Yeah def. Will do this afternoon. Sorry for the delay.

mheilman commented:
oh, no rush!  i just wanted to be clear i was done looking for the time being

beckermr commented:
OK. This is fixed and ready for a second look @mheilman!

mheilman commented:
ah, i just remembered your comment about doing this in chunks.  gonna do a first pass on the whole thing since i'm mostly done.

mheilman commented on muffnn/autoencoder/tests/test_autoencoder.py:
```diff
@@ -0,0 +1,539 @@
+                                   dropout=0.0,
+                                   learning_rate=1e-3)
+
+    assert ae_score_nodropout < ae_score_dropout, ("MSE with dropout should "
```
theoretical question: it seems like this won't be the case in general, for training datasets and nonconvex models.  dropout could actually help avoid some local optimum and end up with a better representation of the training data, no?

beckermr replied:
I am not sure TBH.

beckermr replied:
Empirically I have never seen this happen. You do get better on models on *test* data, but this is just eliminating overfitting.

beckermr commented on Dockerfile:
```diff
@@ -24,14 +24,14 @@ RUN DEBIAN_FRONTEND=noninteractive apt-get update -y && \
-    /bin/bash /Miniconda3-latest-Linux-x86_64.sh -b -p /opt/conda && \
-    rm Miniconda3-latest-Linux-x86_64.sh && \
-    /opt/conda/bin/conda install --yes conda
+    wget --quiet https://repo.continuum.io/miniconda/Miniconda3-4.0.5-Linux-x86_64.sh && \
```
Cool. I think this version had some bug which had not been fixed yet. I will make the change and try it. I used 4.0.5 because I knew it worked.

beckermr replied:
They seemed to have fixed this bug. 4.2.12 works fine.

beckermr commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -0,0 +1,686 @@
+
+        Returns
+        -------
+        Z : numpy array
```
To indicate that this is encoded data. I have seen this notation and using X would be confusing since that is an input.

beckermr commented on muffnn/autoencoder/tests/test_autoencoder.py:
```diff
@@ -0,0 +1,539 @@
+
+    It also handles 0's since they can produce NaNs.
+
+    See scikit.metrics.log_loss for a version with clipping of the
```
Well my original thought was to match the TF computation to get the numerics to be the same. This is probably not happening, but I never went back and removed this.

beckermr replied:
I remember why now. The sklearn one sums over the samples auotmagically. I wanted to test the score of each sample.

beckermr commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -0,0 +1,686 @@
+        # Fan in layers.
+        for i, layer_sz in enumerate(self.hidden_units):
+            # Don't dropout inputs.
+            if i > 0:
```
On second thought, I don't see the issue with only using dropout on the decoding part. The point here is to avoid overfitting and the idea of rescaling inputs (like one-hot encoded stuff), doesn't make sense to me. Using dropout only in the decoding part seems just fine to me.

beckermr commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -0,0 +1,686 @@
+
+                    # This one is painful. TensorFlow does not, AFAIK,
+                    # support assignments to Tensors that come out of
+                    # operations (it does support assignments to Variables).
```
The case condition in `tf.case` is a boolean scalar, so this won't get you around the mask thing.

beckermr commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -0,0 +1,686 @@
+                     binary_msk * tf.nn.sigmoid(t))
+
+            # Anything left uses the default metric.
+            if self._default_indices is not None:
```
Ahh it should be if the default ind set is empty. I will fix.

beckermr commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -0,0 +1,686 @@
+                    diff = isub - tsub
+                    scores += tf.reduce_sum(tf.square(diff),
+                                            reduction_indices=[0])
+                elif self.metric == 'cross-entropy':
```
You could, but then you are doing a bunch of extra work on your data for no reason (multiplying by masks, potentially in efficient gather operations over arrays, etc.). I think the multiple cases make sense here. Also, if you look carefully, the code itself is not duplicated. The code that does the mixed metric case is pretty different since it works on transposed stuff and has `tf.gather` calls.

beckermr commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -0,0 +1,686 @@
+        # Make the dictionary mapping tensor placeholders to input data.
+
+        # Convert sparse inputs to dense.
+        if sp.issparse(X):
```
Two things:

1. No, `check_array` will convert various input sparse formats to a single format, but does not do the sparse to dense conversion. 

2. My goal here was to only convert a sparse input to dense for each mini-batch instead of doing the whole data set at once in order to save memory. This is useful for giant bag-of-words representations, for example.

beckermr commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -0,0 +1,686 @@
+        X = check_array(X, accept_sparse=['csr'])
+
+        random_state = check_random_state(self.random_state)
+        assert self.batch_size > 0, "batch_size <= 0"
```
Ack. This check_random call should only be done once, right?

beckermr replied:
Yep. Moved this inside the is_fitted check.

beckermr commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -0,0 +1,686 @@
+        Parameters
+        ----------
+        X : numpy array or sparse matrix of shape [n_samples, n_features]
+            Encoded data to decode
```
Frack. This doc string is wrong.

beckermr commented on muffnn/autoencoder/tests/test_autoencoder.py:
```diff
@@ -0,0 +1,539 @@
+iris = load_iris()
+
+
+def sigmoid_cross_entropy(ytrue, ypred):
```
Added. I also combined these two functions.

beckermr commented:
@mheilman This is ready for another look!

mheilman commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -141,23 +141,22 @@ def _set_up_graph(self):
 
         # Fan in layers.
         for i, layer_sz in enumerate(self.hidden_units):
-            # Don't dropout inputs.
```
the docstring still says "Note that dropout is not applied to the input layer."

mheilman commented:
hmmm... sorry i didn't think about this until now, but is there a check to make sure that indices aren't specified for both sigmoid and softmax, or for indices above the number of inputs?  input validation like that would be useful to help users avoid silly mistakes

mheilman commented:
The docstring still says "Note that dropout is not applied to the input layer", which isn't the case anymore.

beckermr commented:
Good call on index checking. I will move this into a separate routine.

beckermr commented:
@mheilman I have addressed these comments. Back to you!

mheilman commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -426,36 +425,100 @@ def __getstate__(self):
 
         # Add fitted attributes if the model has been fitted.
         if self._is_fitted:
-            state['input_layer_size_'] = self.input_layer_size_
```
hmm...  i made that comment, but now that i think about it, it's probably more efficient to set the values individually than make a separate dict, but this is fine.  in practice, it won't matter either way here

beckermr replied:
Yeah. We are really talking about microseconds here. :)

mheilman commented on muffnn/autoencoder/autoencoder.py:
```diff
@@ -426,36 +425,100 @@ def __getstate__(self):
+                             " overlap!")
+
+        if not sigmoid_set.isdisjoint(default_set):
+            raise ValueError("Sigmoid indices and default indices cannot"
```
i think this should be an assertion if it shouldn't depend on the user input.

actually, i think it'd be more common to leave out the assertion and instead have unit tests make sure this won't happen.  is testing that the default indices are computed already covered indirectly by the existing tests?

with my previous comment related to this, i was mainly concerned about the user passing in the same indices for softmax and sigmoid, which the check above and ValueError takes care of.

beckermr replied:
Yeah, these extra checks are def covered by the unit tests. I will adjust the code.

mheilman commented:
only a couple minor comments.  thoughts?

beckermr commented:
OK @mheilman. I think we are really close now!

mheilman commented:
```
    __    ______________  ___
   / /   / ____/_  __/  |/  /
  / /   / / __  / / / /|_/ / 
 / /___/ /_/ / / / / /  / /  
/_____/\____/ /_/ /_/  /_/                                
```
(LGTM)

mheilman commented:
my only thought, which would be for another PR, is that it might be nice to have options for penalties on the hidden layer (a la [this paper](http://web.stanford.edu/class/cs294a/sparseAutoencoder.pdf)).
